### PR TITLE
add json enum encoder and decoder

### PR DESCRIPTION
**Pull request by CamClrt**

### CODE REVIEW

File: mergify_engine/json_enum.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# -*- encoding: utf-8 -*-
```
Comment (jd) [inline mergify_engine/json_enum.py:1]: We want to support more than just `Enum` in the future, so that should just be `json` not `json_enum`.

File: mergify_engine/json_enum.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    raise ValueError(f"{v} is an unknown strict merge parameter")
+
+
+class StrictMergeParameterEncoder(json.JSONEncoder):
+    def default(self, v):
+        if isinstance(v, StrictMergeParameter):
+            return {"__enum__": True, "value": v.value}
+        else:
+            return super().default(v)
```
Comment (jd) [inline mergify_engine/json_enum.py:43]: This must work with any `Enum`, not only `StrictMergeParameter`.

File: mergify_engine/json_enum.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    if v == "smart":
+        return StrictMergeParameter.ordered
+    elif v == "smart+fastpath":
+        return StrictMergeParameter.fasttrack
+    else:
+        for name, member in StrictMergeParameter.__members__.items():
+            if v == member.value:
+                return member
+
+    raise ValueError(f"{v} is an unknown strict merge parameter")
```
Comment (jd) [inline mergify_engine/json_enum.py:35]: This should not be there. It has nothing to do with JSON encoding.

File: mergify_engine/tests/unit/test_json_enum.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+with_enum = {
+    "name": "hello",
+    "conditions": [],
+    "actions": {"merge": {"strict": json_enum.StrictMergeParameter.fasttrack}},
+}
+
+
+def test_strict_merge_parameter_encoder():
+    json_file = json.dumps(with_enum, cls=json_enum.StrictMergeParameterEncoder)
+    assert isinstance(json_file, str) is True
```
Comment (jd) [inline mergify_engine/tests/unit/test_json_enum.py:30]: This actually test `json.dumps`, which *always* returns a `str`. So you're not really testing your code but more `json.dumps`.
I think you can just get rid of this test TBH.

File: mergify_engine/tests/unit/test_json_enum.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def test_decode_strict_merge_parameter():
+    json_file = json.dumps(with_enum, cls=json_enum.StrictMergeParameterEncoder)
+    assert (
+        json.loads(json_file, object_hook=json_enum.decode_strict_merge_parameter)
+        == with_enum
+    )
```
Comment (jd) [inline mergify_engine/tests/unit/test_json_enum.py:38]: 👍

**Review by jd — DISMISSED** [review]

File: mergify_engine/json_enum.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        if isinstance(v, StrictMergeParameter):
+            return {"__enum__": True, "value": v.value}
+        else:
+            return super().default(v)
+
+
+def decode_strict_merge_parameter(dct):
+    if "__enum__" in dct:
+        return strict_merge_parameter(dct["value"])
+    return dct
```
Comment (jd) [inline mergify_engine/json_enum.py:49]:
```
def dumps(o):
   return json.dumps(o, cls=…)

def loads(s):
    return json.loads(s, object_hook=decode_strict_merge_parameter)
````

Comment (jd) [reply mergify_engine/json_enum.py:49]: goal:
```
from mergify_engine import json

json.dumps()
```

File: mergify_engine/actions/merge_base.py [hunk 7]
```diff
@@ -54,13 +54,33 @@ class PriorityAliases(enum.Enum):
     high = 3000
 
 
+class StrictMergeParameter(enum.Enum):
+    true = True
+    false = False
+    fasttrack = "smart+fasttrack"
+    ordered = "smart+ordered"
+
+
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:63]: Please don't add this yet, this is not needed in _this_ PR.

File: mergify_engine/actions/merge_base.py [hunk 8]
```diff
@@ -54,13 +54,33 @@ class PriorityAliases(enum.Enum):
...
+    elif v == "smart+fastpath":
+        return StrictMergeParameter.fasttrack
+    else:
+        for name, member in StrictMergeParameter.__members__.items():
+            if v == member.value:
+                return member
+
+    raise ValueError(f"{v} is an unknown strict merge parameter")
+
+
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:83]: ditto

File: mergify_engine/json.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import json
+import enum
+
+from mergify_engine.actions.merge_base import strict_merge_parameter
+
+
+JSON_TYPES = {}
+
+
+def register_types(cls):
```
Comment (jd) [inline mergify_engine/json.py:23]:
```suggestion
def register_type(cls):
```

File: mergify_engine/json.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                "__enum__": True,
+                "class": type(v).__name__,
+                "name": v.name,
+            }
+        else:
+            return super().default(v)
+
+
+def decode_enum(dct):
+    if "StrictMergeParameter" in dct:
```
Comment (jd) [inline mergify_engine/json.py:41]: Why is that here?

File: mergify_engine/json.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import json
+import enum
+
+from mergify_engine.actions.merge_base import strict_merge_parameter
```
Comment (jd) [inline mergify_engine/json.py:17]:
```suggestion
```

File: mergify_engine/json.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def register_types(cls):
+    JSON_TYPES[cls.__name__] = cls
+
+
+class EnumEncoder(json.JSONEncoder):
+    def default(self, v):
+        if isinstance(v, enum.Enum):
+            register_types(type(v))
```
Comment (jd) [inline mergify_engine/json.py:30]: This shouldn't be here. This looks smart, but it's actually error-prone. Take a look at that test:

```python
def test_decode_only():
    json.loads({"__enum__": True, "class": "StrictMergeParameter", "name": "smart"})
```

You think it'll work? No, because `json.dumps` has not been called and the type has not been registered before.

So it's better to make the user explicitely use `register_type` before they can use these JSON functions.

Plus, it's a little bit sub-optimal to call register each time, especially considering this function might be called thousands of times!

File: mergify_engine/json.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            return {
+                "__enum__": True,
```
Comment (jd) [inline mergify_engine/json.py:32]: I'd rather use something like:
```suggestion
                "__pytype__": "enum",
```
As we're likely going to extend this to more than just `Enum` in the future.

File: mergify_engine/tests/unit/test_json.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    "conditions": [],
+    "actions": {"merge": {"strict": Colour.BLUE}},
+}
+
+with_strict_merge_parameter_enum = {
+    "name": "hello",
+    "conditions": [],
+    "actions": {"merge": {"strict": StrictMergeParameter.fasttrack}},
+}
+
```
Comment (jd) [inline mergify_engine/tests/unit/test_json.py:45]: Your `Colour` test is enough.

**Review by jd — DISMISSED** [review]

**Review by jd — DISMISSED** [review]
Hmm you resolved comments but did not fix them. Mistake?

File: mergify_engine/tests/unit/test_json.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+with_enum = {
+    "name": "hello",
+    "conditions": [],
+    "actions": {"merge": {"strict": Colour.BLUE}},
+}
+
+
+def test_decode_enum():
+    json.register_type(type(Colour.BLUE))
```
Comment (jd) [inline mergify_engine/tests/unit/test_json.py:35]:
```suggestion
    json.register_type(Colour)
```

File: mergify_engine/json.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+                "__pytype__": "enum",
+                "class": type(v).__name__,
+                "name": v.name,
+            }
+        else:
+            return super().default(v)
+
+
+def decode_enum(dct):
+    if "enum" in dct.values():
```
Comment (jd) [inline mergify_engine/json.py:38]: This is going to load ALL the values of the dict and check in ALL of them… There's a better way to do this!

File: mergify_engine/json.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        cls_name = dct["class"]
+        for name, member in JSON_TYPES[cls_name].__members__.items():
+            if name == dct["name"]:
```
Comment (jd) [inline mergify_engine/json.py:41]: You're going to iterate over all value of a dict to find a key… Again, there's a better way to do this.

File: mergify_engine/json.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+
+JSON_TYPES = {}
+
+
+def register_type(cls):
+    JSON_TYPES[cls.__name__] = cls
+
+
+class EnumEncoder(json.JSONEncoder):
```
Comment (jd) [inline mergify_engine/json.py:25]:
```suggestion
class JSONEncoder(json.JSONEncoder):
```
This won't support only Enum in the future :)

File: mergify_engine/tests/unit/test_json.py [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    BLUE = 3
+
+
+with_enum = {
+    "name": "hello",
+    "conditions": [],
+    "actions": {"merge": {"strict": Colour.BLUE}},
+}
+
+
```
Comment (jd) [inline mergify_engine/tests/unit/test_json.py:33]: You should also add a test where `dumps` is not used, only `loads`.

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/json.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+                "__pytype__": "enum",
+                "class": type(v).__name__,
+                "name": v.name,
+            }
+        else:
+            return super().default(v)
+
+
+def decode_enum(v):
+    if isinstance(v, dict):
```
Comment (jd) [inline mergify_engine/json.py:38]: I though this was only called when `v` was a dict by `json.loads`? 🤔

File: mergify_engine/tests/unit/test_json.py [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+    "conditions": [],
+    "actions": {"merge": {"strict": Colour.BLUE}},
+}
+
+with_enum_encode = '{"name": "hello", "conditions": [], "actions": {"merge": {"strict": {"__pytype__": "enum", "class": "Colour", "name": "BLUE"}}}}'
+
+
+def test_encode_enum():
+    json.register_type(Colour)
+    assert json.dumps(with_enum) == with_enum_encode
```
Comment (jd) [inline mergify_engine/tests/unit/test_json.py:38]: I think it's fine but be aware this can fail in theory.
A dict is not ordered, so the order in which the strings are printed in the string `with_enum_encode` might change. Which would break the test.

Comment (CamClrt) [reply mergify_engine/tests/unit/test_json.py:38]: if I use `assert data.get("actions").get("merge").get("strict") == Colour.BLUE` is it better ?

File: mergify_engine/json.py (hunk 20 above)
Comment (jd) [reply mergify_engine/json.py:38]: That's actually a question… is that check necessary?

File: mergify_engine/tests/unit/test_json.py (hunk 21 above)
Comment (jd) [reply mergify_engine/tests/unit/test_json.py:38]: what's `data` in your example? 🤔

Comment (CamClrt) [reply mergify_engine/tests/unit/test_json.py:38]: sorry :
```
data = json.loads(json_file)
assert data.get("actions").get("merge").get("strict") == Colour.BLUE
```

Comment (CamClrt) [reply mergify_engine/tests/unit/test_json.py:38]: or I can write it like this:

```
assert (
        json.loads(json_file).data.get("actions").get("merge").get("strict")
        == Colour.BLUE
    )
```

File: mergify_engine/json.py (hunk 20 above)
Comment (CamClrt) [reply mergify_engine/json.py:38]: nop, I have checked the documentation & I have already corrected it in the new PR.

File: mergify_engine/tests/unit/test_json.py (hunk 21 above)
Comment (jd) [reply mergify_engine/tests/unit/test_json.py:38]: If you use `mergify.json.loads` therefore your just writing the same test as `test_decode_enum`.

What you should rather do is use `json.loads` (not the Mergify one) and check the returned data, so:

```python
assert json.loads(json_file) == {"name": "hello", "conditions": [], "actions": {"merge": {"strict": {"__pytype__": "enum", "class": "Colour", "name": "BLUE"}}}}
```

File: mergify_engine/json.py (hunk 20 above)
Comment (jd) [reply mergify_engine/json.py:38]: What new PR? Why isn't that fixed in this PR?

File: mergify_engine/tests/unit/test_json.py [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    "conditions": [],
+    "actions": {
+        "merge": {"strict": {"__pytype__": "enum", "class": "Colour", "name": "BLUE"}}
+    },
+}
+
+
+def test_encode_enum():
+    mergify_json.register_type(Colour)
+    assert mergify_json.dumps(with_enum) == json.dumps(with_enum_encode)
```
Comment (jd) [inline mergify_engine/tests/unit/test_json.py:45]: I'm not sure `json.dumps` is supposedly idempotent so that might work by luck also.
Rather do:
```suggestion
    assert json.loads(mergify_json.dumps(with_enum)) == with_enum_encode
```

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/tests/unit/test_json.py (hunk 22 above)
Comment (CamClrt) [reply mergify_engine/tests/unit/test_json.py]: ok

**Review by jd — APPROVED** [review]
